pyad/reverse_mode.py

```diff
@@ -0,0 +1,167 @@
+import numpy as np
+
+class Reverse:
+
+	def __init__(self, value):
+		self.value = value
+		self.children = []
+		self.grad_value = None
+
+	def get_grad(self):
+		if self.grad_value is None:
+		    self.grad_value = sum(weight * var.get_grad() for weight, var in self.children)
+		return self.grad_value
+
+	def __neg__(self):
+		return self.__mul__(-1)
+
+	def __add__(self, other):
+		if isinstance(other, Reverse):
+			z = Reverse(self.value + other.value)
+			self.children.append((1, z))
+			other.children.append((1, z))
+		else:
+			z = Reverse(self.value + other)
+			self.children.append((1, z))
+		return z
+
+	def __radd__(self, other):
+		return self.__add__(other)
+
+	def __sub__(self, other):
+		sub = other.__neg__()
+		return self.__add__(sub)
+
+	def __rsub__(self, other):
+		sub = self.__neg__()
+		return sub + other
+
+	def __mul__(self, other):
+		if isinstance(other, Reverse):
+			z = Reverse(self.value * other.value)
+			self.children.append((other.value, z))
+			other.children.append((self.value, z))
+		else:
+			z = Reverse(self.value*other)
+			self.children.append((other, z))
+		return z
+
+	def __rmul__(self, other):
+		return self.__mul__(other)
+
+	def __truediv__(self, other):
+		if isinstance(other, Reverse):
+			z = Reverse(self.value / other.value)
+			self.children.append((1/other.value, z))
+			other.children.append((-self.value/other.value**2, z))
+		else:
+			z = Reverse(self.value / other)
+			self.children.append((1/other, z))
+		return z
+
+	def __rtruediv__(self, other):
+		if isinstance(other, Reverse):
+			z = Reverse(other.value / self.value)
+			self.children.append((-other.value/self.value**2, z))
+			other.children.append((1/self.value, z))
+		else:
+			z = Reverse(other / self.value)
+			self.children.append((-other/self.value**2, z))
+		return z
+
+	def __pow__(self, other):
+		if isinstance(other, Reverse):
+			z = Reverse(self.value**other.value)
+			self.children.append((other.value*self.value**(other.value-1), z))
+			other.children.append((self.value**other.value*np.log(self.value), z))
+		else:
+			z = Reverse(self.value**other)
+			self.children.append((other*self.value**(other-1), z))
+		return z
+
+	def __rpow__(self, other):
+		z = Reverse(other**self.value)
+		self.children.append((other**self.value*np.log(other), z))
+		return z
+
+# Elementary functions
+def _elementary_op(obj, fn, deriv_fn):
+	if isinstance(obj, Reverse):
```

Comment:
So that its return type is consistent, this function should always return a `Tensor` object. So I think the conversion thing I discussed above should also apply here as well